storehaus-kafka/src/main/scala/com/twitter/storehaus/kafka/JavaFutureToTwitterFutureConverter.scala

```diff
@@ -24,60 +24,82 @@ import com.twitter.util.{Future, Try, Promise}
 import scala.annotation.tailrec
 
 /**
-  * Utility class for converting Java futures to Twitter's 
-  * @param waitTimeMs Time spent sleeping by the thread converting java futures to
+  * Utility class for converting Java futures to Twitter's
+ *
+ * @param waitTimeMs Time spent sleeping by the thread converting java futures to
   *                   twitter futures when there are no futures to convert in ms
   */
 private[kafka] class JavaFutureToTwitterFutureConverter(waitTimeMs: Long = 1000L) {
-  
-  def apply[T](javaFuture: JFuture[T]): Future[T] = {
-    val promise = new Promise[T]()
-    poll(Link(javaFuture, promise))
-    promise
+
+  case class Link[T](future: JFuture[T], promise: Promise[T]) {
+    def maybeUpdate: Boolean = future.isDone && {
+      promise.update(Try(future.get()))
+      true
+    }
+
+    def cancel(): Unit = {
+      promise.setException(new Exception("Promise not completed"))
+      future.cancel(true)
+    }
   }
-  
+
+  sealed abstract class State
+  case object Closed extends State
+  case class Open(links: List[Link[_]]) extends State
+  val EmptyState: State = Open(Nil)
+
   private val pollRun = new Runnable {
     override def run(): Unit =
       try {
-        while (!Thread.currentThread().isInterrupted)
-          loop(list.getAndSet(Nil))
+        if (!Thread.currentThread().isInterrupted)
+          loop(list.getAndSet(EmptyState))
       } catch {
         case e: InterruptedException =>
       }
-    
+
     @tailrec
-    def loop(links: List[Link[_]]): Unit = {
-      val notDone = links.filterNot(_.maybeUpdate)
-      if (links.isEmpty || notDone.nonEmpty) Thread.sleep(waitTimeMs)
-      loop(list.getAndSet(notDone))
+    def swapOpen(old: List[Link[_]]): (List[Link[_]], Boolean) = list.get match {
+      case Closed => (old, false)
+      case s@Open(links) =>
+        if (list.compareAndSet(s, Open(old))) (links, true)
+        else swapOpen(links)
+    }
+
+    @tailrec
+    def loop(state: State): Unit = state match {
+      case s@Open(links) =>
+        val notDone = links.filterNot(_.maybeUpdate)
+        if (links.isEmpty || notDone.nonEmpty) Thread.sleep(waitTimeMs)
+        loop(list.getAndSet(Open(notDone)))
```

Comment:
I think this needs to be `swapOpen` and if you get false, you need to `foreach(_.cancel)`, right?
```scala
swapOpen(notDone) match {
  case (cancels, true) => cancels.foreach(_.cancel)
  case (next, false) => loop(Open(next))
}
```